packages/vm/src/instance.rs

```diff
@@ -114,6 +111,31 @@ where
                 }),
             },
         };
+
+        let rw_imports = imports! {
+            "env" => {
+                // Writes the given value into the database entry at the given key.
+                // Ownership of both input and output pointer is not transferred to the host.
+                // Returns 0 on success. Returns negative value on error.
+                "write_db" => Func::new(move |ctx: &mut Ctx, key_ptr: u32, value_ptr: u32| -> i32 {
+                    do_write::<S, Q>(ctx, key_ptr, value_ptr)
+                }),
+                // Removes the value at the given key. Different than writing &[] as future
+                // scans will not find this key.
+                // Ownership of both key pointer is not transferred to the host.
+                // Returns 0 on success. Returns negative value on error.
+                "remove_db" => Func::new(move |ctx: &mut Ctx, key_ptr: u32| -> i32 {
+                    do_remove::<S, Q>(ctx, key_ptr)
+                }),
+            },
+        };
+
+        let mut import_obj = imports! { || { setup_context::<S, Q>() }, "env" => {}, };
```

Comment:
I think this approach will work even better for iterator than for DB writes, because `write_db` and `remove_db` will be required imports of a contract, even if not used for query. So we'll need to convert this PR to stubs.

Reply:
Do you want to Pull out a PR just doing iterator properly (using the next `extend` approach)? Then build this on top, maybe using stubs.

One way of avoiding a separate cache for query and writable is to store the `allow_db_writes` flag in the context. (Since you are right, we do need to expose the imports for linking). Then we can update it everytime we instantiate the contract (one more arg after gas limit).

Reply:
> Do you want to Pull out a PR just doing iterator properly (using the next `extend` approach)? Then build this on top, maybe using stubs.

Jap, let's start with the easy one.